cluster/raft_query_endpoints.go

```diff
@@ -284,31 +284,30 @@ func (s *Raft) Query(ctx context.Context, req *cmd.QueryRequest) (*cmd.QueryResp
 		))
 	defer t.ObserveDuration()
 
-	resp := &cmd.QueryResponse{}
-	err := backoff.Retry(func() error {
-		var err error
-
-		if s.store.IsLeader() {
-			resp, err = s.store.Query(req)
-			return err
-		}
+	if s.store.IsLeader() {
+		return s.store.Query(req)
+	}
 
-		leader := s.store.Leader()
-		if leader == "" {
-			err = s.leaderErr()
+	// find out who the leader is
+	var leader string
+	f := func() error {
+		if leader = s.store.Leader(); leader == "" {
+			err := s.leaderErr()
 			s.log.Warnf("query: could not find leader: %s", err)
 			return err
 		}
 
-		resp, err = s.cl.Query(ctx, leader, req)
-		if err != nil {
-			s.log.WithField("leader", leader).Errorf("query: failed to query leader: %s", err)
-			// Don't retry if the actual query fails
-			return backoff.Permanent(err)
-		}
 		return nil
-		// Retry at most for 2 seconds, it shouldn't take longer for an election to take place
-	}, backoff.WithContext(backoff.WithMaxRetries(backoff.NewConstantBackOff(200*time.Millisecond), 10), ctx))
+	}
```

Comment:
**suggestion:**  could we put this function inside the `backoff.Retry()` call in L303 for better readability

Reply:
Hi @moogacs!

Thank you for the feedback! Moving f inside `retry` would make the code look like this 
```
	var leader string
	bf := backoff.WithMaxRetries(backoff.NewConstantBackOff(200*time.Millisecond), 10)
	if err := backoff.Retry(func() error {
		if leader = s.store.Leader(); leader == "" {
			err := s.leaderErr()
			s.log.Warnf("query: could not find leader: %s", err)
			return err
		}

		return nil
	}, backoff.WithContext(bf, ctx)); err != nil {
		s.log.Errorf("query: failed to find leader after retries: %s", err)
		return &cmd.QueryResponse{}, err
	}
```
The code is not easy to read, as f would be nested with retry(), which could make it more difficult to distinguish between what is being retried (f) and retry logic itself. 

I believe the current structure
```
	var leader string
	f := func() error {
		if leader = s.store.Leader(); leader == "" {
			err := s.leaderErr()
			s.log.Warnf("query: could not find leader: %s", err)
			return err
		}

		return nil
	}
	bf := backoff.WithMaxRetries(backoff.NewConstantBackOff(200*time.Millisecond), 10)
	if err := backoff.Retry(, backoff.WithContext(bf, ctx)); err != nil {
		s.log.Errorf("query: failed to find leader after retries: %s", err)
		return &cmd.QueryResponse{}, err
	}
```
offers better clarity, but, I'm open to making adjustments if you feel it's necessary. Please let e know you thoughts!

Reply:
Hello Redouan!

I disagree I think the first form is easier to read as the whole code block is retried and indented that way in the call to `Retry` itself. With the `f := ...` form, it makes it harder to understand what f does when you see the call to `Retry` . We have this pattern in a few other places and we're trying to move away from it.

Also you should be able to nest the `backoff.WithMaxRetries` and `backoff.WithContext` together in the call line in the call to `Retry` this avoids an additional line before the call to `Retry` which can be confusing.

Reply:
Hey Loic!

Thanks for the follow-up! understand your point, and while I still believe the current approach is clearer, I've implemented the changes to align with the pattern you 're using. I've also nested backoff.* as suggested.
Let me know if there is anything else! 🙂